### ISSUE TITLE
Detail about the 11 iteration during training

### ISSUE THREAD

**Issue by Nastu-Ho** [facebookresearch/segment-anything#689]
During the training process, each batch of image embeddings will be iteratively input to the mask decoder 11 times, and the mask decoder will output mask logits 11 times. I wonder if the mask logits of each output need to be used to calculate the loss, or is it just the last prediction to get the mask logits to calculate the loss?

**Comment by Vishawjeet-rmsl** [comment]
I'm also wondering what the authors mean by this. Is it for the mask generation for SA-1B or during the training of SAM itself. i.e., if someone is trying to train a SAM model (with different architecture) do they have to implement this as well.

**Comment by filipeas** [comment]
> During the training process, each batch of image embeddings will be iteratively input to the mask decoder 11 times, and the mask decoder will output mask logits 11 times. I wonder if the mask logits of each output need to be used to calculate the loss, or is it just the last prediction to get the mask logits to calculate the loss?

From what is apparently described in the paper (in the training session in the appendix), the process follows an update rule ``` L = 1/B * sum((1/64 * sum((1/11 * sum(Lb,m,i))))) ```.

In other words, for each iteration with prompts (8 or 11, depending on the attempts made in the paper), you calculate the logit loss (choosing the best of the three hypotheses generated by the model. Read the training session in the appendix; they explain how to calculate this), accumulate it, and take the average (here's a personal question, as it's not clear whether you should average the iterations or apply opt.optimizer() per iteration. My interpretation leads me to believe the average should be calculated). Then, you take this average and apply opt.optimizer() (here again, it's important to note the description of how the training is done, as the paper mentions the random selection of 64 masks) for each mask present in the image (here, I'm assuming you'll be using the SA-1B, where each image has 100 possible masks).

That's basically it.
